src/metabase/query_processor/middleware/update_used_cards.clj

```diff
@@ -0,0 +1,21 @@
+(ns metabase.query-processor.middleware.update-used-cards
+  (:require
+   [clojure.string :as str]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.query-processor.store :as qp.store]
+   [metabase.util.log :as log]
+   #_{:clj-kondo/ignore [:discouraged-namespace]}
+   [toucan2.core :as t2]))
+
+(defn update-used-cards!
+  "Around middleware that update last_used_at of all cards that were used during a query execution."
+  [qp]
+  (fn [query rff]
+    (letfn [(rff* [metadata]
+              ;; TODO: should this be done async for perf purposes? maybe this and process-userland-query should share a
+              ;; pool?
+              (let [used-cards (lib.metadata/cached-cards (qp.store/metadata-provider))]
```

Comment:
You should probably check that `used-cards` isn't empty before doing this

Reply:
added a seq check